### ISSUE TITLE
Investigate Why the LLM Is Not Learning Gender Information

### ISSUE THREAD

**Issue by tanzir5** [odissei-lifecourse/life-sequencing-dutch#134]
##  Context
We have noticed that the LLM fails to learn or retain gender information. A simple experiment where the life-course embedding is passed to a logistic regression model to classify the gender. The gender token is part of the life-sequence, so it is reasonable to expect the embeddings to have retained this information. 

Initial hypotheses suggest potential information loss due to the following factors:
1. **Averaging Token Embeddings**: This might result in the loss of crucial contextual information, such as gender-specific terms. Everything is getting diluted. 
2. **Using Static Embeddings**: Freezing BERT layers may prevent the model from adapting to the gender classification task.
3. **Including Padding Embeddings in Averages**: Averaging embeddings without excluding padding tokens introduces noise, potentially zeroing out information. Related issue --  #133

## Experiment with original BERT
To isolate and understand the issue, we will test a pretrained (on English) BERT model on a proxy task: predicting whether a text contains the token “walk” or “run.” This simple task will help us validate the impact of various settings on performance. The reason behind doing this with original BERT is a lot of the stuff we are doing is application-oriented, as opposed to having a concrete theoretical background. It is unfortunately true for a lot of ML stuff. 

When our model is not working as expected, there can be two reasons:
1. We made a dumb mistake. 
2. The setting does not work in practice, even though the theory seems sound.

When the original BERT works on something but our pop2vec model does not, this tells us it is probably the first thing, not the second.  

### Steps:
1. **Download the Data**:
   - Use a subset of the Wikipedia dataset.
2. **Model**:
   - Use a pretrained BERT model.
3. **Fine-Tuning Settings**:
   - **Embedding Extraction**:
     - `[CLS]` token embedding.
     - Averaging token embeddings (ensure padding is excluded).
   - **Prediction Layer**:
     - A linear layer with one neuron + sigmoid + BCELoss (mimics a logistic regression model).
     - A slightly more complex setting with two layers mimicking Hugging Face fine-tuning.
   - **Freezing BERT Layers**:
     - Use static embeddings (no fine-tuning).
     - End-to-end fine-tuning (all layers trainable).

This results in **8 experimental settings**:
- 2 embedding types × 2 prediction layers × 2 freezing settings.

The setting with mean embedding + one linear layer + static (no fine-tuning) is a replication of what Ana is doing right now.
## Action Items
- [x] Fix #133 
- [x] Implement the experiments using each of the 8 fine-tuning settings.
- [x] Compare model performance across settings to identify the root cause of information loss.
- [x] Document results and propose adjustments based on findings.

**Comment by tanzir5** [comment]
The following table contains the experiment results for BERT using wikipedia dataset for detecting if "man" or "woman" token was present in the first 64 tokens of an input wiki article. 

| Finetune LLM | Embedding | Layers | Acc |
|--------------|-----------|--------|-----|
| No           | Mean      | 1      | 89  |
| Yes          | Mean      | 1      | 99  |
| No           | CLS       | 1      | 83  |
| Yes          | CLS       | 1      | 99  |
| No           | Mean      | 2      | 88  |
| Yes          | Mean      | 2      | 99  |
| No           | CLS       | 2      | 78  |
| Yes          | CLS       | 2      | -   |


Summary:
1. Finetuning can make a big difference for even simple tasks like "find if a token is present in the input"
2. Mean embedding is in general better than CLS for simple tasks if we are not fine-tuning. I think I finally understand why CLS embedding does as good or even better for classification when we are fine-tuning. RoBERTa (the LLM Facebook built following BERT) argues there is no point in having the CLS pre-training objectives. If we removed CLS objectives, the CLS token will learn very little during pre-training. But it will learn a lot during fine-tuning since it is essentially a weighted sum of embeddings from all tokens for a specific task and its own weight naturally will be very small since its initial embedding means nothing. 

"man" vs "woman" in the first 64 tokens of a wikipedia article has the additional advantage that other tokens are influenced by the "man" and "woman" (e.g., "man" will have tokens like "he", "him", whereas "woman" will have tokens like "she", "her", etc). 

For a more difficult task like "little" vs "big", the performance improves from 72% using static embeddings to 99% accuracy using fine-tuned embeddings. 


Conclusion:
Our model has also learnt similarly to dfifferentiate between "male" and "female" gender tokens with  78-81% accuracy using mean embeddings. Specifically, BERT with mean embedding with one layer without fine-tuning got 83% accuracy differentiating "little" vs "big". We with the same setting got 78% differentiating "male" vs "female" gender token. 

So the results are okay I believe. If the results do not improve after fine-tuning, then there is reason to worry.

**Comment by tanzir5** [comment]
This notebook contains the code:
https://github.com/odissei-lifecourse/life-sequencing-dutch/blob/main/notebooks/run_or_walk.ipynb

**Comment by tanzir5** [comment]
For the same task with "small" vs "little" with one layer prediction setting,

| Finetune LLM | Embedding | Acc% |
|--------------|-----------|------|
| No           | Mean      | 83   |
| Yes          | Mean      | 99   |
| No           | CLS       | 73   |
| Yes          | CLS       | 99   |
